# Heart disease analysis
Simple Classification workfow - 

This workflow assumes data is ready to be used with machine learning models (is numerical and has no missing values.) 

In [69]:
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd 
import numpy as np
import sklearn

In [17]:
#Importing the dataset
heart_disease = pd.read_csv("./data/heart-disease.csv")

#Viewing the data
heart_disease.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Will use all the columns except the target column to predict the target column.
Using a patient's medical and demographic data to predict whether or not they have heart-disease

In [3]:
# Creating X (all the feature coliumns)
X = heart_disease.drop("target", axis = 1)

# Creating Y (the target column)
Y = heart_disease["target"]

In [49]:
# Splitting the data into training and test sets 
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X,Y)

# Viewing the data shapes
X_train.shape, X_test.shape, Y_train.shape, Y_test.shape

((227, 13), (76, 13), (227,), (76,))

# Choosing the Model
Since, we are working on a classification problem, we will use the RandomForestClassifier model which is part of the ensemble module

In [68]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier()

# Fitting the model to the data to make a prediction

The model will attempt to learn the patterns in a dataset by calling the fit() function on it and passing the data 

In [72]:
model.fit(X_train,Y_train)

RandomForestClassifier()

Since the model has learned patterns in data, we can use them to make a prediction with the predict() function

In [9]:
# Making Predictions
Y_preds = model.predict(X_test)

In [10]:
# Viewing the predictions (This will be in the same format as Y_test)
Y_preds

array([0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0,
       0, 1, 0, 1, 0, 0, 1, 1, 1, 0], dtype=int64)

# Randomly selecting a data set
choosing - (dataset[210])

In [53]:
X_test.loc[210]

age          57.0
sex           1.0
cp            2.0
trestbps    128.0
chol        229.0
fbs           0.0
restecg       0.0
thalach     150.0
exang         0.0
oldpeak       0.4
slope         1.0
ca            1.0
thal          3.0
Name: 210, dtype: float64

In [54]:
heart_disease.loc[210]

age          57.0
sex           1.0
cp            2.0
trestbps    128.0
chol        229.0
fbs           0.0
restecg       0.0
thalach     150.0
exang         0.0
oldpeak       0.4
slope         1.0
ca            1.0
thal          3.0
target        0.0
Name: 210, dtype: float64

In [55]:
# Making a prediction on a single sample (has to be array)
model.predict(np.array(X_test.loc[210]).reshape(1,-1))

array([0], dtype=int64)

## Evaluating the model

To evaluate a trained model by calling the score() function and passing it a collection of data 

In [58]:
# On the training data set
model.score(X_train, Y_train) * 100

84.58149779735683

In [59]:
# On the test set (UNSEEN)
model.score(X_test, Y_test) * 100

78.94736842105263

## Experiments to improve (Hyperparameter Tuning)
Trying to improve the model's predictions with hyperparameter tuning

In [74]:
# Trying different numbers of estimators (n_estimators is a hyperparameter we can change)
np.random.seed(12)
for i in range(10, 100, 10):
    print(f"Tried model with {i} estimators...")
    model = RandomForestClassifier(n_estimators=i).fit(X_train, Y_train)
    print(f"Model accuracy on test set: {model.score(X_test, Y_test)}")
    print("")

Tried model with 10 estimators...
Model accuracy on test set: 0.8289473684210527

Tried model with 20 estimators...
Model accuracy on test set: 0.75

Tried model with 30 estimators...
Model accuracy on test set: 0.7763157894736842

Tried model with 40 estimators...
Model accuracy on test set: 0.8157894736842105

Tried model with 50 estimators...
Model accuracy on test set: 0.8157894736842105

Tried model with 60 estimators...
Model accuracy on test set: 0.8157894736842105

Tried model with 70 estimators...
Model accuracy on test set: 0.8157894736842105

Tried model with 80 estimators...
Model accuracy on test set: 0.8157894736842105

Tried model with 90 estimators...
Model accuracy on test set: 0.7763157894736842



## Conclusion :
So the model best works with 10 estimators with an accuracy of -> 0.8289473684210527

Testing different hyperparmeters with a validation set or cross validation:

In [78]:
from sklearn.model_selection import cross_val_score

# Trying different numbers of estimators with cross-validation and no cross-validation
np.random.seed(12)
for i in range(10, 100, 10):
    print(f"Trying model with {i} estimators....")
    model = RandomForestClassifier(n_estimators = i).fit(X_train, Y_train)
    print(f"Model accuracy on test set : {model.score(X_test, Y_test) * 100}%")
    print(f"Cross-validation score: {np.mean(cross_val_score(model, X, Y, cv = 5)) * 100}%")
    print("")

Trying model with 10 estimators....
Model accuracy on test set : 82.89473684210526%
Cross-validation score: 79.53551912568307%

Trying model with 20 estimators....
Model accuracy on test set : 78.94736842105263%
Cross-validation score: 80.5191256830601%

Trying model with 30 estimators....
Model accuracy on test set : 80.26315789473685%
Cross-validation score: 82.83606557377048%

Trying model with 40 estimators....
Model accuracy on test set : 78.94736842105263%
Cross-validation score: 82.49726775956286%

Trying model with 50 estimators....
Model accuracy on test set : 80.26315789473685%
Cross-validation score: 80.5191256830601%

Trying model with 60 estimators....
Model accuracy on test set : 81.57894736842105%
Cross-validation score: 82.15300546448088%

Trying model with 70 estimators....
Model accuracy on test set : 80.26315789473685%
Cross-validation score: 82.50273224043717%

Trying model with 80 estimators....
Model accuracy on test set : 80.26315789473685%
Cross-validation score

## Saving the model
To export and save the trained model to use it later using Python's pickle module.

In [79]:
import pickle

# Saving trained model in a file
pickle.dump(model, open("Heart-disease-classification-model.pkl", "wb"))

# Loading a saved model


In [85]:
loaded_model = pickle.load(open("Heart-disease-classification-model.pkl", "rb"))

# Making a prediction on a single model
loaded_model.predict(np.array(X_test.loc[210]).reshape(1,-1))

array([1], dtype=int64)